File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        a, r = 0, len(nums) - 1
        while a <= r:
            m = a  + ((r - a) // 2)
            if nums[m] > target:
                r = m - 1 
            elif nums[m] < target:
                a = m + 1
            else:
                return m
        return -1

File: Data Structures & Algorithms/three-integer-sum/submission-1.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        sort(nums.begin(), nums.end());
        int n = nums.size();

        for (int i = 0; i < n - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue; // skip dup i

            int j = i + 1, k = n - 1;
            while (j < k) {
                int total = nums[i] + nums[j] + nums[k];
                if (total < 0) {
                    j++;
                } else if (total > 0) {
                    k--;
                } else {
                    result.push_back({nums[i], nums[j], nums[k]});
                    j++;
                    k--;
                    while (j < k && nums[j] == nums[j - 1]) j++; // skip dup j
                    while (j < k && nums[k] == nums[k + 1]) k--; // skip dup k
                }
            }
        }
        return result;
    }
};
